File: go/0417-pacific-atlantic-water-flow.go
func pacificAtlantic(heights [][]int) [][]int {
    ROWS, COLS := len(heights), len(heights[0])
    pac, atl := make(map[int]bool), make(map[int] bool)
    
    var dfs func(int, int, map[int]bool, int)
    dfs = func(r, c int, visit map[int]bool, prevHeight int) {
        if (
            visit[r*COLS + c] ||
            r < 0 ||
            c < 0 ||
            r == ROWS ||
            c == COLS ||
            heights[r][c] < prevHeight) {
            return;
        }
        visit[r*COLS + c] = true
        dfs(r + 1, c, visit, heights[r][c])
        dfs(r - 1, c, visit, heights[r][c])
        dfs(r, c + 1, visit, heights[r][c])
        dfs(r, c - 1, visit, heights[r][c])
    }
    
    for c := 0; c < COLS; c++ {
        dfs(0, c, pac, heights[0][c])
        dfs(ROWS - 1, c, atl, heights[ROWS - 1][c])
    }
    
    for r := 0; r < ROWS; r++ {
        dfs(r, 0, pac, heights[r][0])
        dfs(r, COLS - 1, atl, heights[r][COLS - 1])
    }
    
    res := make([][]int, 0)
    for r := 0; r < ROWS; r++ {
        for c := 0; c < COLS; c++ {
            if pac[r*COLS + c] && atl[r*COLS + c] {
                res = append(res, []int{r, c})
            }
        }
    }
    return res
}

File: swift/0417-pacific-atlantic-water-flow.swift
class Solution {
    func pacificAtlantic(_ heights: [[Int]]) -> [[Int]] {
        let rows = heights.count 
        let cols = heights[0].count
        var pac = Set<[Int]>()
        var atl = Set<[Int]>()
        var res = [[Int]]()

        func dfs(_ r: Int, _ c: Int, _ visit: inout Set<[Int]>, _ prevHeight: Int) {
            if r < 0 || c < 0 || r == rows || c == cols || visit.contains([r, c]) || heights[r][c] < prevHeight {
                return
            }
            visit.insert([r, c])
            dfs(r + 1, c, &visit, heights[r][c])
            dfs(r - 1, c, &visit, heights[r][c])
            dfs(r, c + 1, &visit, heights[r][c])
            dfs(r, c - 1, &visit, heights[r][c])
        }

        for c in 0..<cols {
            dfs(0, c, &pac, heights[0][c])
            dfs(rows - 1, c, &atl, heights[rows - 1][c])
        }

        for r in 0..<rows {
            dfs(r, 0, &pac, heights[r][0])
            dfs(r, cols - 1, &atl, heights[r][cols - 1])
        }

        for r in 0..<rows {
            for c in 0..<cols {
                if pac.contains([r, c]) && atl.contains([r, c]) {
                    res.append([r, c])
                }
            }
        }

        return res
    }
}